garak/resources/beast/beast_attack.py

```diff
@@ -0,0 +1,284 @@
+from garak.generators import Generator
+from garak.resources.common import rejection_strings, advbench_path
+
+beast_resource_data = garak._config.transient.basedir / "resources" / "beast" / "data"
+
+
+def format_chat(prompt: str):
+    # TODO: Implement this functionality!
+    return prompt
+
+
+@torch.no_grad()
+def evaluate(generator, prompt, best_candidate):
+    result = False
+    best_candidate_str = generator.tokenizer.decode(best_candidate)
+    input_str = prompt + best_candidate_str
+    outputs = generator.generate(input_str)
+    results = check_jailbreak(outputs)
+    if all(results):
+        result = True
+    return result, outputs[0]
+
+
+@torch.no_grad()
+def evaluate_target(generator, prompt, best_candidate, target):
+    result = False
+    best_candidate_str = generator.tokenizer.decode(best_candidate)
+    input_str = prompt + best_candidate_str
+    outputs = generator.generate(input_str)
+    for output in outputs:
+        if target in output:
+            result = True
+            return result, output
+    return result, outputs[0]
+
+
+def tensorize(generator: Generator,
+              prompt: str,
+              response: str,
+              suffix_ids: Union[list[int], torch.Tensor, None] = None) -> tuple[torch.Tensor, int]:
+    """
+    Turn prompt and response into tensors of tokens
+    Args:
+        generator: targeted generator
+        prompt: prompt to the model
+        response: response from the model
+        suffix_ids: suffix_ids, if any.
+    Returns:
+        encoding_tensor: Tensor of tokenized inputs
+        split: Index of tensor where model output begins
+    """
+    suffix_tensor = torch.tensor([], dtype=torch.long)
+    if suffix_ids is not None:
+        suffix_tensor = torch.tensor(suffix_ids, dtype=torch.long) if isinstance(suffix_ids, list) else suffix_ids
+    prompt_encoding = generator.tokenizer.encode(prompt, add_special_tokens=False, return_tensors="pt").squeeze(0)
+    output_encoding = generator.tokenizer.encode(response, add_special_tokens=False, return_tensors="pt").squeeze(0)
+    encoding_tensor = torch.cat((prompt_encoding, suffix_tensor, output_encoding))
+    split = encoding_tensor.shape[0] - output_encoding.shape[0]
+
+    return encoding_tensor, split
+
+
+@torch.no_grad()
+def get_perplexity(generator: Generator,
+                   prompt: str,
+                   response: str,
+                   suffix_ids: Union[list[int], torch.Tensor, None] = None) -> float:
+    """
+    Given a generator and a prompt, returns the perplexity of the input.
+    Args:
+        generator: targeted generator
+        prompt: prompt to the model
+        response: response from the model
+        suffix_ids: suffix_ids, if any.
+
+    Returns:
+        perplexity: float value for the perplexity of the prompt.
+    """
+    encoding_tensor, split = tensorize(generator, prompt, response, suffix_ids)
+    input_tensor = encoding_tensor.unsqueeze(0).to(generator.model.device)
```

Comment:
This whole class looks highly coupled to `huggingface.Model` should we restrict this to that class type for now or is there some way to abstract away details like which device the model is using?

Reply:
Unfortunately, it very much is highly coupled to `huggingface` generators. Would love to extend support. We do check for the relevant subclasses and methods early on, but it would definitely be ideal to support more.